### ISSUE TITLE
Add more contextual information to ParserException

### ISSUE THREAD

**Issue by tobiasrodehutskors** [jparsec/jparsec#43]
I want to use JParsec to parse commands received at a command line interface. Possible commands might look like this:

> list -d <ID>

or

> runs -x <ID>

Now we want to display error messages to the command line user, if he messes something up. For example, if he enters "list -d ", a possible error message might look like this:

> "You need to specify an ID for the option '-d' of the command 'list'."

However ParseErrorDetails does not provide enough contextual information for this.  We only have the information that a 'decimal integer' was expected, but an 'EOF' received and the position of the parsing error. Unfortunately, this does not allow to distinguish between the different possible commands as the exact same ParseErrorDetails object might indicate an error in the list command as well as an error in the runs command.

Is there any possibility to receive more contextual information from an ParserException? I think it would be great if we were able to attach arbitrary objects to a ParserException. This could be archived by adding a special Parser:

``` java
public class SpecialExceptionParser<T> extends Parser<T> {

        Parser<T> parser;
        Object errorObject;

        SpecialExceptionParser(Parser<T> parser, Object errorObject) {
            this.parser = parser;
            this.errorObject = errorObject;
        }

        @Override
        public T parse(CharSequence source, String moduleName, SourceLocator sourceLocator) {
            try{
                super.parse(source, moduleName, sourceLocator);
            } catch(ParserException e) {
                throw ExtendedParserException(e, errorObject);
            }
        }

    }
```

This would allow to encapsulate all Parsers of special interrest and attach an arbitary object to identify what went wrong:

``` java
SpecialExceptionParser(Scanners.string("te"), arbitraryObject);
```

Do you have any idea how more contextual information can be added to the ParserException?

**Comment by fluentfuture** [comment]
Hi Tobias.

Technically, you could use label
<http://javadox.com/jparsec/jparsec/2.0.1/org/codehaus/jparsec/Parser.html#label(java.lang.String)>()
to give a logical name to the syntactical element, that can *somewhat*
customize the error message. But it won't give the exact error message you
want.

I think it would be a lot more flexible if you take the two-pass approach
like many other programming language parsers. That is: the parser is simply
to turn text into AST, and then the type checker/validator validates the
AST according to semantic rules.

In this case, if each flag with an optional value were parsed into a Flag
object, with name/value pair, then you can apply any semantic rule and
validation, throw any exception that's appropriate.

In other words, keep the parser dumb and concentrated on its core task:
parsing text. And leave validation to the second step once parser is done
with its work. One bonus: you can then display multiple errors at the same
time. There are INDEX
<http://javadox.com/jparsec/jparsec/2.0.1/org/codehaus/jparsec/Parsers.html#INDEX>
parser that can pass the source location into the AST object, so that the
semantic validation can point back into the original text.

Hope that makes sense.

**Comment by abailly** [comment]
@fluentfuture I am all for an AST-based approach but a potential drawback is that this might entail changing your syntax to accommodate for "user errors". In the example given by @tobiasrodehutskors this means changing the parser for `-d <ID>` from something like `token(D_FLAG).followedBy(idParser())` into   `token(D_FLAG).followedBy(idParser().optional())` which might not make sense at the syntax level.

**Comment by fluentfuture** [comment]
I'm sympathetic to the desire of a more programmatically usable
ParserException. I recall at one point in the early versions of jparsec, I
used to have something called "ExecutionTrace " that'll report the stack of
production rules that failed. It didn't prove to serve any real purpose so
it was removed later.

Jumping the gun a bit:

I can see us adding something similar with good use cases. Still, I worry
that the catch() block could still be messy since it needs to first see
that it's the "idParser()" that failed, and then one level up tell that
it's the D_FLAG.followedBy(idParser()) that failed.

A more robust solution may be to use label() on the trace so that all the
caller has to do is to walk up the trace to find the expected label. That
way the parser do:

D_FLAG.followedBy(idParser()).label("d_flag");


And the catch block can do:

try {
  ..

} catch (ParserException e) {

  for (ProductionRule rule : e.getProductionRuleStack()) {

    if ("d_flag".equals(rule.getLabel())) {

      // report error.

    }

  }

}



How does that sound?

That said, I'm still unsure how far it can get us though. At some point you
really have exceeded the expressiveness power of ParserException and have
to use AST for validation at a higher level abstraction.



Side note: I'm actually used to "-d" being optional in commandline tools.
Like "ls -l". :)
